rfcs/0051-hard-fork-package-generation.md

```diff
@@ -0,0 +1,148 @@
+
+In the future, we will add additional jobs to the pipeline that will also perform automated tests against the release candidate.
+
+### Testing the Hard Fork Package
+
+In the future, we wish to automate the process of testing a hard fork release candidate. However for the upcoming hard fork, we will instead rely on a manual testing process in order to verify that the hard fork release candidate is of sufficient quality to release. The goal of this testing at this layer is to determine that the new package is configured to the proper genesis state, and that a network can be initialized using the package. We do not perform wider scope functional test suites at this stage, as the version of the software being built when deploying a hard fork will have already been tested thoroughly; we are only interested in testing workflows introduced or affected by the hard fork package generation process.
+
+Ideally, we deploy a very short testnet with community members involved to verify the hard fork package. However, this is not possible to do without modifying the ledger, unless we were to involve the larger Mina staking community. Instead, we will test the hard fork package by initializing seed nodes and checking some very specific criteria. This is considered ok as our goal here is not to test networking, but the initialization states of a daemon when it hits the genesis timestamp, and the state we read from it once it has fully initialized the genesis chain.
+
+The test plan for the seed node would go as follows:
+
+1. Run a seed node with a custom runtime config that sets the genesis timestamp to be 10 minutes in the future.
+2. Monitor the seed node logs to ensure that it boots up and begins sleeping until genesis timestamp.
+3. Once the seed has passed the genesis timestamp, monitor logs to ensure the node goes through bootstrap only to pass through into participation with the genesis block as its best tip.
+4. Run a script against the node that will sample random accounts from each of the genesis ledgers and compare the data served by the daemon against the data in the genesis state we captured from the old chain.
+
+These tests will ensure that the hard fork package is setup properly to initialize the new chain with the correct state.
+
+In order to enable these tests, we will need to ensure we have GraphQL or CLI endpoints available for each of the ledgers (snarked ledger, staged ledger, and the epoch ledgers). Then we will need to write a script that reads the ledger files input to the hard fork package generation timeline and performs the sampling test described in step (4).
+
+<!-- ISSUE: this won't work because we can't modify the ledger for block production
```

Comment:
We could turn [this check](https://github.com/MinaProtocol/mina/blob/e4e96c1892da6d2a2a22d1df34acb084da5e27ff/src/lib/consensus/proof_of_stake.ml#L761) off based on a value in the proof-config. Then, any pk can generate a block based on the stake in any existing account. This would be nice for this stage of testing, but also potentially for an automated nightly hard-fork test.

Reply:
That's an option. I get worried with these kind of backdoors though. We need to be careful in how it is designed so that we can be sure that no one ever takes advantage of it. The main attack vector here is that an adversary could try and add some sneaky code to our system that somehow configures this insecurely for an actual release of the software.

Reply:
They don't need to be particularly sneaky about it either. Anyone can checkout the released source code, modify it to their liking, compile and  connect it to the network to do nefarious things.